src/NHibernate/Linq/ReWriters/ConditionalQueryReferenceExpander.cs

```diff
@@ -0,0 +1,195 @@
+				_queryType = queryType;
+			}
+			
+			protected override Expression VisitQuerySourceReference(QuerySourceReferenceExpression expression)
+			{
+				if (_queryType.IsAssignableFrom(expression.Type))
+				{
+					_queryReferenceCount++;
+				}
+
+				return base.VisitQuerySourceReference(expression);
+			}
+
+			public static int CountReferences(Expression node)
+			{
+				var visitor = new QueryReferenceCounter(node.Type);
+				visitor.Visit(node);
+				return visitor._queryReferenceCount;
+			}
+		}
+
+		private abstract class ConditionalQueryReferenceExpressionRewriter<T, TVisitor> : RelinqExpressionVisitor
+			where T : Expression
+			where TVisitor : ConditionalQueryReferenceExpressionRewriter<T, TVisitor>, new()
+		{
+			protected System.Type QueryType { get; private set; }
+			protected T OuterExpr { get; private set; }
+			private bool _skipUpdate;
+
+			protected override Expression VisitBinary(BinaryExpression node)
+			{
+				if (node.NodeType != ExpressionType.Coalesce)
+				{
+					return base.VisitBinary(node);
+				}
+
+				// Coalesce expressions must be rewritten to conditionals to keep their logical meaning
+				// (x ?? y).Prop --> x != null ? x.Prop : y.Prop
+				return Expression.Condition(
+					Expression.NotEqual(node.Left, Expression.Constant(null, node.Left.Type)),
+					this.Visit(node.Left),
+					this.Visit(node.Right));
+			}
+
+			protected override Expression VisitConditional(ConditionalExpression node)
+			{
+				_skipUpdate = true;
+				var test = this.Visit(node.Test);
+				_skipUpdate = false;
+				return Expression.Condition(test, this.Visit(node.IfTrue), this.Visit(node.IfFalse));
+			}
+
+			protected override Expression VisitQuerySourceReference(QuerySourceReferenceExpression expression)
+			{
+				if (!_skipUpdate && QueryType.IsAssignableFrom(expression.Type))
+				{
+					return Rewrite(expression);
+				}
+
+				return base.VisitQuerySourceReference(expression);
+			}
+
+			protected abstract T Rewrite(QuerySourceReferenceExpression expression);
+
+			public static Expression Rewrite(Expression expression, T outerExpr)
+			{
+				var visitor = new TVisitor { OuterExpr = outerExpr, QueryType = expression.Type };
+				return visitor.Visit(expression);
+			}
+		}
+
+		private class ConditionalQueryReferenceMemberExpressionRewriter : ConditionalQueryReferenceExpressionRewriter<MemberExpression, ConditionalQueryReferenceMemberExpressionRewriter>
+		{
+			protected override MemberExpression Rewrite(QuerySourceReferenceExpression expression)
+			{
+				return Expression.MakeMemberAccess(expression, OuterExpr.Member);
+			}
+		}
+
+		private class ConditionalQueryReferenceMethodCallExpressionRewriter : ConditionalQueryReferenceExpressionRewriter<MethodCallExpression, ConditionalQueryReferenceMethodCallExpressionRewriter>
+		{
+			protected override MethodCallExpression Rewrite(QuerySourceReferenceExpression expression)
+			{
+				return Expression.Call(expression, OuterExpr.Method, OuterExpr.Arguments);
```

Comment:
I have tried by adding a `string SomeMethod()` on the `Client` entity, registered with:

```c#
public class ClientGenerator : BaseHqlGeneratorForMethod
{
	public ClientGenerator()
	{
		SupportedMethods = new[]
		{
			ReflectHelper.GetMethod<Client>(
				c => c.SomeMethod())
		};
	}

	public override HqlTreeNode BuildHql(
		MethodInfo method,
		Expression targetObject,
		ReadOnlyCollection<Expression> arguments,
		HqlTreeBuilder treeBuilder,
		NHibernate.Linq.Visitors.IHqlExpressionVisitor visitor)
	{
		return treeBuilder.Dot(
			visitor.Visit(targetObject).AsExpression(),
			treeBuilder.Ident("Name"));
	}
}

public class ClientHqlGeneratorsRegistry : DefaultLinqToHqlGeneratorsRegistry
{
	public ClientHqlGeneratorsRegistry()
	{
		this.Merge(new ClientGenerator());
	}
}
```
And:
```c#
protected override void Configure(Configuration configuration)
{
	configuration.SetProperty(
		Environment.LinqToHqlGeneratorsRegistry,
		typeof(ClientHqlGeneratorsRegistry).AssemblyQualifiedName);
}

...

[Test]
public void WhereClauseMethodCall()
{
	AreEqual(
		// Actual
		q => q.Where(i => (i.Project.BillingClient ?? i.Project.Client ?? i.Client).SomeMethod().StartsWith("A")),
		// Expected
		q => q.Where(i => (i.Project.BillingClient != null ? i.Project.BillingClient.SomeMethod() : i.Project.Client != null ? i.Project.Client.SomeMethod() : i.Client.SomeMethod()).StartsWith("A"))
	);
}
```

It fails, because it seems the `Client` entities are not considered as being query source references by ReLinq when no properties are accessed. So the method call case cannot work, unless adding some `VisitMember` overrides. But this would likely catch cases which should not be rewritten.

I now also wonder, would this PR correctly ignore non-leaf nodes? I mean, if the entity model allows things like `Client.Client`, will it correctly rewrite only the last `Client`?

Reply:
I've fixed the tests (and the join logic).

Not sure what you mean in your last comment.
By the time the conditional/coalesce expansion runs for non-subqueryt expressions the joins have already been generated so property chains have all been reduced to `[joinref].Property`.

Reply:
Ok, it is likely just me which is still not enough knowledgeable about how ReLinq, expression trees and the NH Linq provider work.